kernel/ipi.cc

```diff
@@ -90,3 +90,53 @@ on_ipicall()
     }
   }
 }
+
+std::atomic<int> paused_cpu_counter {0};
+enum pause_state { Ready, Pausing, Paused, Unpausing, Unpaused };
+std::atomic<enum pause_state> pause_state {Ready};
+enum pause_state READY = Ready;
+
+// Use IPI to pause other cores until update is complete.
+void
+pause_other_cpus(void)
+{
+  while (!atomic_compare_exchange_strong(&pause_state, &READY, Pausing)); // wait for any previously paused CPUs to resume
```

Comment:
Shouldn't this be using one of the member functions on `atomic<...>`? Based on skimming [this page](https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange) it looks like the weak compare and exchange version is the one we want.

```cpp
while(pause_state.atomic_compare_exchange_weak(Ready, Pausing)) {
    nop_pause();
}
```